client/client_metrics.go

```diff
@@ -0,0 +1,93 @@
+package client
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	namespaceHttpClient = "httpclient"
+
+	metricNameRequestDurationSeconds = "request_duration_seconds"
+	metricNameRequestTotal           = "request_total"
+
+	labelNameUrl    = "url"
+	labelNameMethod = "method"
+	labelNameStatus = "status"
+
+	labelValueErr = "error"
+)
+
+type httpClientMetrics struct {
+	durationSeconds *prometheus.HistogramVec
+	requestTotal    *prometheus.CounterVec
+}
+
+func newHttpClientMetrics(constLabels prometheus.Labels) *httpClientMetrics {
+	m := &httpClientMetrics{
+		durationSeconds: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace:   namespaceHttpClient,
+			Name:        metricNameRequestDurationSeconds,
+			Help:        "Histogram of duration of outgoing http requests",
+			ConstLabels: constLabels,
+		}, []string{labelNameUrl, labelNameMethod}),
+		requestTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace:   namespaceHttpClient,
+			Name:        metricNameRequestTotal,
+			Help:        "Count of total outgoing http requests, with its result status in labels",
+			ConstLabels: constLabels,
+		}, []string{labelNameUrl, labelNameMethod, labelNameStatus}),
+	}
+
+	return m
+}
+
+func (metric *httpClientMetrics) observeDuration(req *http.Request, startTime time.Time) {
+	url := getHttpReqMetricUrl(req)
+	method := req.Method
+
+	metric.durationSeconds.WithLabelValues(url, method).Observe(time.Since(startTime).Seconds())
+}
+
+func (metric *httpClientMetrics) observeResult(req *http.Request, resp *http.Response, err error) {
+	url := getHttpReqMetricUrl(req)
+	method := req.Method
+	status := getHttpRespMetricStatus(resp, err)
+
+	metric.requestTotal.WithLabelValues(url, method, status).Inc()
+}
+
+// Describe implements prometheus.Collector interface
+func (metric *httpClientMetrics) Describe(descs chan<- *prometheus.Desc) {
+	metric.durationSeconds.Describe(descs)
+	metric.requestTotal.Describe(descs)
+}
+
+// Collect implements prometheus.Collector interface
+func (metric *httpClientMetrics) Collect(metrics chan<- prometheus.Metric) {
+	metric.durationSeconds.Collect(metrics)
+	metric.requestTotal.Collect(metrics)
+}
+
+func getHttpReqMetricUrl(req *http.Request) string {
```

Comment:
IMO get route URL is too broad

I intentionally make the function name specific, because this function is returning a "partial" URL that is not commonly needed. It will only return the hostname and the path of a URL, i.e. if full URL is http://www.example.com/abc?param1=a&param2=b, this function returns only http://www.example.com/abc (more example in the [test](https://github.com/trustwallet/go-libs/pull/172/files#diff-5117a7ff1203a62d1aede10c77034b41b9ae636a0d179c9cb330aba77ac2874bR135))

This is required because storing metrics with high cardinality labels is [very costly](https://prometheus.io/docs/practices/naming/#:~:text=Remember%20that%20every,sets%20of%20values.)